src/main/java/io/mosip/iiitb/odte/lib/MessageBrokerWrapper.java

```diff
@@ -0,0 +1,146 @@
+    private final OnDemandTemplateExtractionConsumerImpl odteConsumer;
+    private final KafkaConsumer<String, String> consumer;
+    private final RSACryptoTool rsaCryptoTool;
+    private final Logger logger;
+
+    @Inject
+    public MessageBrokerWrapper(
+            OnDemandAppConfig mbConfig,
+            OnDemandTemplateExtractionConsumerImpl odteConsumer,
+            RSACryptoTool rsaCryptoTool,
+            Logger logger
+    ) {
+        this.logger = logger;
+        String topic = mbConfig.kafkaTopic();
+
+        this.consumer = initializeKafka(
+                mbConfig
+        );
+        DummyConsumerRebalanceListener rebalancedListener = new DummyConsumerRebalanceListener();
+        this.consumer.subscribe(Collections.singletonList(topic), rebalancedListener);
+        this.odteConsumer = odteConsumer;
+        this.rsaCryptoTool = rsaCryptoTool;
+    }
+    public void start() {
+        try {
+            loopIt();
+        } finally {
+            this.consumer.close();
+        }
+    }
+
+    private KafkaConsumer<String, String> initializeKafka(
+            OnDemandAppConfig config
+    ) {
+        Properties props = new Properties();
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, config.kafkaBrokers());
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, config.kafkaGroupId());
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        props.put(
+                ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+                StringDeserializer.class.getName()
+        );
+        props.put(
+                ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+                StringDeserializer.class.getName()
+        );
+
+        logger.debug(
+                String.format(
+                        "Listening on %s:%s\nTopics = %s\nGroupId = %s\nBrokers = %s\n",
+                        config.kafkaHostname(),
+                        config.kafkaPort(),
+                        config.kafkaTopic(),
+                        config.kafkaGroupId(),
+                        config.kafkaBrokers()
+                )
+        );
+        return new KafkaConsumer<>(props);
+    }
+
+    private void loopIt() {
+        logger.debug("Started listening to events");
+        while (true) {
+            final ConsumerRecords<String, String> consumerRecords =
+                    this.consumer.poll(Duration.ofMillis(100));
+            if (consumerRecords.count() == 0) {
+                continue;
+            }
+
+            for (ConsumerRecord<String, String> record : consumerRecords) {
+                DecryptedOnDemandTemplateRecord drc = parseOnDemandTemplateRecord(record);
+                if (drc != null) {
+                    EventConsumerResponse evtConsumerResponse = this.odteConsumer.processRecord(drc);
+                    logger.debug("Processed a record and got response = {}", evtConsumerResponse);
+                }
+            }
+            this.consumer.commitAsync();
```

Comment:
You should limit the number of records that you will read from kafka.
If there is sudden 10k records received in one poll, since the logic is sequential, you may take a while before doing the next poll and the kafka will rebalance.

Reply:
Done, now only 24 records will be processed in a single poll, this value is configurable through the "mosip.iiitb.ondemand.kafka.maxRecordsToProcessAtOnce" property